### ISSUE TITLE
MsalClientException "No refresh token was found"

### ISSUE THREAD

**Issue by 4gus71n** [AzureAD/microsoft-authentication-library-for-android#570]
Hello,

I'm using the **com.microsoft.identity.client:msal:0.2.2** version of the MSAL library.

**Context:**

My Android App uses an MVVM architecture, the views (Activities, Fragments) communicate with the ViewModel, the ViewModel is injected with a bunch of Interactors using Dagger, each Interactor has a bunch of Repositories, and the Repositories are the ones making the API calls to the backend using Retrofit. All my Repositories extend from a BaseRepository class, which uses a MSAL wrapper called **MSALSessionProvider** to fetch the auth token that I need to authenticate my requests against the backend. This **MSALSessionProvider** internally uses the **PublicClientApplication** to fetch the token through the **PublicClientApplication#acquireTokenSilentAsync()** method. For sake of testing I have a QA Tools dashboard where I can change some QA-related configurations. One of this settings is to always send the **forceRefresh** parameter from the **PublicClientApplication#acquireTokenSilentAsync()** method, as **true**. This is just to enforce some stress over the App and make sure that the entire App is going to work properly when it actually needs to re-fetch the auth token from the backend.

**Issue:**

I noticed one error that is super tricky to reproduce. This is the stacktrace that I got from Crashlytics:

![image](https://user-images.githubusercontent.com/2055468/55977376-10833f80-5c65-11e9-95cd-5fc4dd4c1a81.png)

To reproduce this error I need to have this QA flag enabled (to refetch the token from the backend all the time) and then open the App, close it from the recent list, open it again, close it, and so on. Some times just by doing this 6 times I was able to reproduce the issue, some other times It took me like 20 times of doing this.

**Repercussions:**

As I mentioned before, all my Repository classes, that are the ones in charge of sending the requests to the backend are hooked to the MSAL wrapper, which internally uses the MSAL library stuff. Always, before making any request I use the **PublicClientApplication#acquireTokenSilentAsync()**  to fetch the auth token. If this fails the Repository is not going to try to hit the backend. Thus, when this error happens my App doesn't fetch data from the backend.

**Question:**

I just wanna know what's the context of this issue to see if I can patch it somehow. Is there any specific reason for this issue? Cache corruption? Something off on my **IAccount** instance?

Thanks

**Comment by shoatman** [comment]
@iambmelt - I'm curious how IAccount was returned if there was no corresponding refresh token.

**Comment by 4gus71n** [comment]
@shoatman @iambmelt Thank you guys for taking some time to look into this. About the IAccount, the only outstanding thing that I have on my code is the **super nasty bugfix** for this [issue](https://github.com/AzureAD/microsoft-authentication-library-for-android/issues/495#issuecomment-473436019)

Maybe it's related? BTW do you guys know if there's some ETA for the new API [changes](https://github.com/AzureAD/microsoft-authentication-library-for-android/issues/495#issuecomment-454204440) I think that maybe that would fix both issues.

**Comment by iambmelt** [comment]
@4gus71n This workaround would _absolutely_ cause a loss of SSO state, leading to issue you're seeing above. Clearing the cache's contents deletes the refresh token we use to reauthorize the account.

If you have a steady-repro of the `getAccounts()` issue, that's what we should try to better understand. Any info you can provide on this front would be helpful.

**Comment by 4gus71n** [comment]
@iambmelt Okay, sorry I'm totally on the dark about the other issue. Only thing that I can say is that I get the exact same stack trace as the person that posted the issue. And trying to fetch the `IAccount` to call the `aquireTokenSilentAsync()` is one of the first things that I do when I open the App. I've some sort of Splash screen, and I do a validation there to see if the user is logged in or not.

All my testing I do it on Samsung phones (Samsung J7 and J6) since they're more prone to errors.

I think I'll wait for the new API fix, again, no ETA on this?

**Comment by iambmelt** [comment]
@4gus71n The changes around `getAccounts()` isn't _really_ going to resolve this issue - the `getAccounts` issue, as I understand it so far, seems to be related to undecryptable data in the cache. So far, my assumption is that this is the result of [unreliable keystore behavior](https://doridori.github.io/android-security-the-forgetful-keystore/).

If you're able to repro this bad behavior regularly, I am _extremely intersted_ in understanding how it happens and what we can do to minimize it

**Comment by iambmelt** [comment]
@4gus71n I committed this change recently to plaster-over this issue - it's a far-cry from ideal.

https://github.com/AzureAD/microsoft-authentication-library-common-for-android/issues/480
https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/481

**Comment by 4gus71n** [comment]
I know that this is an old closed thread, but I think I found out some useful information about this issue. I just finished implementing what I was talking about on this issue https://github.com/AzureAD/microsoft-authentication-library-for-android/issues/847. 

At first, everything was working perfectly, but then I started getting the old good `MsalClientException: No refresh token was found. ` exceptions

Since I added code into an Interceptor to synchronously fetch the auth token, I think that the issue may come as a result of trying to get the auth token from different threads, asynchronously.

This is the code that I have on the Interceptor:

![image](https://user-images.githubusercontent.com/2055468/70169702-0e656980-16aa-11ea-9345-552132514d99.png)

Then on that `getMsalAuthTokenBearer()` function I have this:

![image](https://user-images.githubusercontent.com/2055468/70169749-2dfc9200-16aa-11ea-8c61-9778ab27a48d.png)

That `alwaysRefresh` refresh variable, it's a boolean variable that I use for QA purposes. To test that the app is going to work fine, even if I'm trying to fetch a new auth token all the time. When I switch that boolean variable on is that I start seeing the issue.

Think that in my case, as soon as I open the app, I hit 5 different endpoints, all those API calls are going to go through the interceptor and try to fetch a brand new auth token.

I don't have any clear solution so far, but I'm going to try to cache the auth token and keep track of when it expires.

Any feedback, tip would be appreciated. But I thought about sharing this information so It may shed some light on this issue.

@iambmelt @shoatman

**Comment by iambmelt** [comment]
> I know that this is an old closed thread, but I think I found out some useful information about this issue. I just finished implementing what I was talking about on this issue #847.
> 
> At first, everything was working perfectly, but then I started getting the old good `MsalClientException: No refresh token was found. ` exceptions
> 
> Since I added code into an Interceptor to synchronously fetch the auth token, I think that the issue may come as a result of trying to get the auth token from different threads, asynchronously.
> 
> This is the code that I have on the Interceptor:
> 
> ![image](https://user-images.githubusercontent.com/2055468/70169702-0e656980-16aa-11ea-9345-552132514d99.png)
> 
> Then on that `getMsalAuthTokenBearer()` function I have this:
> 
> ![image](https://user-images.githubusercontent.com/2055468/70169749-2dfc9200-16aa-11ea-8c61-9778ab27a48d.png)
> 
> That `alwaysRefresh` refresh variable, it's a boolean variable that I use for QA purposes. To test that the app is going to work fine, even if I'm trying to fetch a new auth token all the time. When I switch that boolean variable on is that I start seeing the issue.
> 
> Think that in my case, as soon as I open the app, I hit 5 different endpoints, all those API calls are going to go through the interceptor and try to fetch a brand new auth token.
> 
> I don't have any clear solution so far, but I'm going to try to cache the auth token and keep track of when it expires.
> 
> Any feedback, tip would be appreciated. But I thought about sharing this information so It may shed some light on this issue.
> 
> @iambmelt @shoatman

@4gus71n Are you using MSAL 1.0.0 and still having this issue?

**Comment by 4gus71n** [comment]
@iambmelt Yes, actually I updated to the latest release (1.0.0-hf1) this morning, just in case. But I was using 1.0.0 before that.

**Comment by iambmelt** [comment]
@4gus71n In that case, I am going to reopen this issue.

To quickly summarize, when you encounter this issue, it seems to happen when you acquire tokens for multiple resource simultaneously, have I understood correctly?

**Comment by 4gus71n** [comment]
If by "multiple resources" you mean my backend's endpoints, then **yes**. What happens is that:

1) I set the `acquireTokenSilent()` to force refresh through the `alwaysRefresh` variable:

![image](https://user-images.githubusercontent.com/2055468/70175831-2fcc5280-16b6-11ea-993d-0a417f48df76.png)

2) I open the app, the app is going to hit different endpoints, my app has a ViewPager with different fragments for different sections of the app. I'm going to trigger 5 different API calls to my backend, each of those API calls are going to go through my Interceptor, which is going to try to fetch the authentication token through the synchronous API.

3) I'm going to get the exception.

Notice that this only happens if I force the PublicClientApp to refresh through the `forceRefresh(Boolean)` function.

![image](https://user-images.githubusercontent.com/2055468/70176132-ad905e00-16b6-11ea-89b0-fdc8486d8925.png)

**Comment by iambmelt** [comment]
@4gus71n Ok, more questions. When you perform your `acquireToken` request, are you using `withOtherScopesToAuthorize` and including all of your services/scopes in that first request?

For context, MSAL maintains a single refresh token for a user; using `force_refresh` _forces_ MSAL to hit AAD and acquire a new one. In this case, we have multiple threads making a request essentially simultaneously, there could be some contention/race condition for this token as it's being written/read from the cache.

I suspect this is what we're seeing; it would also explain why some requests intermittently cannot find the RT; if it has been deleted by a prior request, for example

**Comment by iambmelt** [comment]
@4gus71n Additionally, could you include a stack trace from the latest version of the library where you have a repro?

**Comment by 4gus71n** [comment]
@iambmelt Thanks! More than happy to answer any questions:

>When you perform your acquireToken request, are you using withOtherScopesToAuthorize and including all of your services/scopes in that first request?

I don't use anything like that, I just use a scope `Array<String>` and that's it. I'm not using the `acquireToken` to prompt the MSAL WebView login screen, I'm using the `signIn` function:

![image](https://user-images.githubusercontent.com/2055468/70178394-f9dd9d00-16ba-11ea-8e48-9a1623e31e02.png)

(That's code from the wrapper where I keep the PublicClientApplication instance)

From there on, I just use the `acquireTokenSilent()` I never actually call the `acquireToken()` on my app.

**Comment by 4gus71n** [comment]
@iambmelt Also, here's a [gist](https://gist.github.com/4gus71n/2fb0ff5f57117d60617c9fa51f4e36a7) with the stacktrace. Let me know if that's useful enough.

**Comment by iambmelt** [comment]
> @iambmelt Also, here's a [gist](https://gist.github.com/4gus71n/2fb0ff5f57117d60617c9fa51f4e36a7) with the stacktrace. Let me know if that's useful enough.

ACK - thanks @4gus71n

**Comment by nfgallimore** [comment]
Experiencing same problem

**Comment by alinciupeiu** [comment]
I am receiving the same error (Failed to retrieve SSO token) when trying to call application.acquireTokenSilent(scopes, account, authority) on three different threads at the same time. Do you have any updates on this issue?

**Comment by nfgallimore** [comment]
@alinciupeiu I have been calling acquireTokenSilentAsync on three different threads at the same time no problems

**Comment by alinciupeiu** [comment]
@nfgallimore 
I am using Broker (Microsoft Authenticator) Authentication.

Most of the time it works ok, even when calling the aquireTokenSilent(scopes, account, authority) from different threads in the same time.

However, from time to time, when I re-launch the application and perform a lot of requests in background in parallel, I call the acquireTokenSilent() for each request, I see this error in logs:
2020-03-09 14:01:58.932 11876-12019/com.iquestgroup.meetingroom.store W/com.microsoft.identity.common.internal.result.MsalBrokerResultAdapter:  [2020-03-09 12:01:58 - {"thread_id":"32394","correlation_id":"26c2620e-ea6b-467d-ad24-be4f5fea7c54"}] Received a Service exception from Broker : no_tokens_found Android 27
2020-03-09 14:01:58.935 11876-12019/com.iquestgroup.meetingroom.store W/com.microsoft.identity.common.internal.result.MsalBrokerResultAdapter:  [2020-03-09 12:01:58 - {"thread_id":"32394","correlation_id":"26c2620e-ea6b-467d-ad24-be4f5fea7c54"}] Unable to parse json Android 27
2020-03-09 14:01:58.939 11876-12019/com.iquestgroup.meetingroom.store E/BrokerAuthServiceStrategy:acquireTokenSilentWithAuthService:  [2020-03-09 12:01:58 - {"thread_id":"32394","correlation_id":"26c2620e-ea6b-467d-ad24-be4f5fea7c54"}] No refresh token was found.  Android 27
    com.microsoft.identity.common.exception.ServiceException: No refresh token was found. 
        at com.microsoft.identity.common.internal.result.MsalBrokerResultAdapter.getServiceException(MsalBrokerResultAdapter.java:283)
        at com.microsoft.identity.common.internal.result.MsalBrokerResultAdapter.getBaseExceptionFromBundle(MsalBrokerResultAdapter.java:226)
        at com.microsoft.identity.common.internal.result.MsalBrokerResultAdapter.getAcquireTokenResultFromResultBundle(MsalBrokerResultAdapter.java:360)
        at com.microsoft.identity.client.internal.controllers.BrokerAuthServiceStrategy$3.perform(BrokerAuthServiceStrategy.java:204)
        at com.microsoft.identity.client.internal.controllers.BrokerAuthServiceStrategy$3.perform(BrokerAuthServiceStrategy.java:200)
        at com.microsoft.identity.client.internal.controllers.BrokerAuthServiceStrategy.performAuthServiceOperation(BrokerAuthServiceStrategy.java:119)
        at com.microsoft.identity.client.internal.controllers.BrokerAuthServiceStrategy.acquireTokenSilent(BrokerAuthServiceStrategy.java:199)
        at com.microsoft.identity.client.internal.controllers.BrokerMsalController$2.perform(BrokerMsalController.java:313)
        at com.microsoft.identity.client.internal.controllers.BrokerMsalController$2.perform(BrokerMsalController.java:309)
        at com.microsoft.identity.client.internal.controllers.BrokerMsalController.invokeBrokerOperation(BrokerMsalController.java:196)
        at com.microsoft.identity.client.internal.controllers.BrokerMsalController.acquireTokenSilent(BrokerMsalController.java:308)
        at com.microsoft.identity.common.internal.controllers.TokenCommand.execute(TokenCommand.java:73)
        at com.microsoft.identity.common.internal.controllers.TokenCommand.execute(TokenCommand.java:42)
        at com.microsoft.identity.common.internal.controllers.CommandDispatcher.executeCommand(CommandDispatcher.java:138)
        at com.microsoft.identity.common.internal.controllers.CommandDispatcher.access$200(CommandDispatcher.java:53)
        at com.microsoft.identity.common.internal.controllers.CommandDispatcher$1.run(CommandDispatcher.java:95)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
        at java.lang.Thread.run(Thread.java:764)

**Comment by z3ntu** [comment]
I believe I'm seeing the same issue with msal 1.5.1:
```
     Caused by: com.microsoft.identity.client.exception.MsalClientException: No refresh token was found. 
        at com.microsoft.identity.client.internal.controllers.MsalExceptionAdapter.msalExceptionFromBaseException(MsalExceptionAdapter.java:51)
        at com.microsoft.identity.client.SingleAccountPublicClientApplication$3.onError(SingleAccountPublicClientApplication.java:339)
        at com.microsoft.identity.client.SingleAccountPublicClientApplication$3.onError(SingleAccountPublicClientApplication.java:325) 
        at com.microsoft.identity.common.internal.controllers.CommandDispatcher$3.run(CommandDispatcher.java:209) 
        at android.os.Handler.handleCallback(Handler.java:883) 
        at android.os.Handler.dispatchMessage(Handler.java:100) 
        at android.os.Looper.loop(Looper.java:214) 
        at android.app.ActivityThread.main(ActivityThread.java:7682) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:516) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:950) 
     Caused by: com.microsoft.identity.common.exception.ClientException: No refresh token was found. 
        at com.microsoft.identity.client.internal.controllers.LocalMSALController.acquireTokenSilent(LocalMSALController.java:319)
        at com.microsoft.identity.common.internal.commands.SilentTokenCommand.execute(SilentTokenCommand.java:75)
        at com.microsoft.identity.common.internal.commands.SilentTokenCommand.execute(SilentTokenCommand.java:41)
        at com.microsoft.identity.common.internal.controllers.CommandDispatcher.executeCommand(CommandDispatcher.java:166)
        at com.microsoft.identity.common.internal.controllers.CommandDispatcher.access$200(CommandDispatcher.java:65)
        at com.microsoft.identity.common.internal.controllers.CommandDispatcher$2.run(CommandDispatcher.java:126)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:919)
```

**Comment by iambmelt** [comment]
Documenting a know _expected_ case where this can happen before I close this issue.

When a refresh token is invalidated due to expiring or revocation (such as when a user changes their account password) AAD will issue a special error code to the client library with the message `bad_token`; this code signals to the client that this token cannot be used again for any purpose and will _always_ be rejected.

When the client library sees this message it deletes the `bad_token` from the cache such that subsequent requests will not attempt to use it. Once this occurs, the caller should cease to attempt silent token requests and instead direct the user to perform interactive (UI-based) authentication such that they reenter their username/password

The message "No refresh token was found" may occur under any of the following conditions:
- Attempting to silent-auth an account that has not signed-in or has been removed via FLW mode
- Attempting to silent-auth an account whose RT has been revoked/expired and the client received a `bad_token` error message from AAD
- Attempting to silent-auth an account with does not correspond to the current B2C policy -- please note that account from PolicyA cannot be used for PolicyB. Accounts are exclusive to a policy
- If the RT on disk is unavailable due cache corruption, encryption key rotation or otherwise

As I am stepping back from being an active contributor to the project, please consider opening a new issue if you see this error message and the above guidance doesn't provide a satisfactory explanation for the behavior you're seeing.

As a additional tip, consider seeking help via StackOverflow and inspecting the contents of your app's logs; if posting logs publicly please ensure you're careful to avoid posting personally/organizationally identifiable data:
https://docs.microsoft.com/en-us/azure/active-directory/develop/msal-logging-android
